Add reshape_as_blocks function

larrybradley opened the pull request:
This is often a handy function to have.  Using this function, I refactored `block_reduce` to remove astropy's dependency on `scikit-image`.  An additional bonus is `block_reduce` is faster now.

I split the `block_*` functions from `nddata/utils.py` into a new module called `nddata/blocks.py`.
The first few commits are necessary git gymnastics to preserve the history of both files.

pllim commented:
Technically, this didn't make the 4.1 feature freeze for new PRs last Friday but I'll defer to @bsipocz .

larrybradley commented:
@bsipocz noted that the `matplotlib` docs dependency had been being pulled in by `scikit-image`.  I explicitly added `matplotlib` as a docs dependency.

bsipocz commented:
@pllim - we haven't branched, so whatever is merged this week will be certainly in (as I'm not willing to mess with the history for the sake of forcing a double deadline)

astrofrog commented on CHANGES.rst:
```diff
@@ -79,6 +79,8 @@ astropy.nddata
   when specifying the bit flags to be used or ignored when converting a bit
   field to a boolean. [#10095]
 
+- Added ``reshape_as_blocks`` function. [#10214]
```
Can you give more context here as to what it does?

larrybradley replied:
I added some more context.  It does exactly what it says on the tin.  :smile:

larrybradley commented:
Because I split the `block_*` functions from `utils.py` into a new `blocks.py` module, importing these functions from `nddata.utils` would no longer work:  e.g., `from astropy.nddata.utils import block_reduce` would fail.  I just added a commit that allows the old imports from `nddata.utils` to still work, but issues a deprecation warning about the new module.  The module-level `__class__` reassignment can be replaced with PEP562 when the minimum required python version is 3.7.

bsipocz commented:
I think v4.2 will have minimun python 3.7, so a cleanup PR can be opened once we're done with 4.1

bsipocz commented:
(maybe not right away, as keeping the 3.6 tests around for a bit will significantly help with the 4.1.1 bugfix release)

pllim commented on astropy/nddata/tests/test_blocks.py:
```diff
@@ -0,0 +1,164 @@
+from astropy.nddata import reshape_as_blocks, block_reduce, block_replicate
+
+
+class TestReshapeAsBlocks:
```
Add a test to check that `reshaped` is really a view and not a copy?

larrybradley replied:
Done.

pllim commented on astropy/nddata/blocks.py:
```diff
@@ -0,0 +1,206 @@
+
+def reshape_as_blocks(data, block_size):
+    """
+    Reshape a data array into blocks.
```
This needs a "version added" directive?

larrybradley replied:
I can add that, but I didn't think astropy used it very much.

pllim replied:
Its usage is not consistent throughout the package, but doesn't hurt?

larrybradley replied:
Done.

bsipocz replied:
>  I didn't think astropy used it very much

not yet, but hopefully we change that 😅

larrybradley commented on astropy/nddata/blocks.py:
```diff
@@ -0,0 +1,206 @@
+        raise ValueError('block_size must be a scalar or have the same '
+                         'length as the number of data dimensions')
+
+    if np.any(block_size <= 0):
```
Yes, done.

larrybradley commented on astropy/nddata/blocks.py:
```diff
@@ -0,0 +1,206 @@
+           [0.  , 0.  , 0.25, 0.25],
+           [0.5 , 0.5 , 0.75, 0.75],
+           [0.5 , 0.5 , 0.75, 0.75]])
+
```
I actually wanted the blank line to separate the two examples.

pllim replied:
Hmm, as long as it renders fine, I guess it's ok.

larrybradley replied:
It does.

pllim commented:
I'll wait for the docs to render and have a look before approving, otherwise LGTM. Thanks!

pllim commented on astropy/nddata/blocks.py:
```diff
@@ -0,0 +1,208 @@
+    >>> import numpy as np
+    >>> from astropy.nddata import block_reduce
+    >>> data = np.arange(16).reshape(4, 4)
+    >>> block_reduce(data, 2)    # doctest: +SKIP
```
Wait... why is it still skipped but my comment marked as resolved? Hope you can clarify. Thanks.

larrybradley replied:
Hmm, I accepted your suggestion above.

larrybradley replied:
I just fixed that (again).

pllim commented:
Please don't merge yet. I am gonna have one last look, since Larry did ask me to review. 😉

larrybradley commented:
Here are some timings comparing astropy 4.0 (old) and this PR (new).  For 1000x1000 arrays, the new version is ~2x faster.  For 5000x5000 arrays, the new version is ~4-10x faster (resulting in ~0.12 s speedup).

```python
>>> import numpy as np
>>> from astropy.nddata import block_reduce
>>> np.random.seed(123)
>>> data = np.random.random((1000, 1000))
>>> %timeit block_reduce(data, (10, 10))
old: 2.30 ms    new: 1.48 ms    ratio: 1.55
>>> %timeit block_reduce(data, (50, 50))
old: 1.39 ms    new: 669 µs    ratio: 2.08
>>> %timeit block_reduce(data, (100, 100))
old: 1.32 ms    new: 580 µs    ratio: 2.28

>>> data = np.random.random((5000, 5000))
>>> %timeit block_reduce(data, (10, 10))
old: 160 ms    new: 37.6 ms    ratio: 4.26
>>> %timeit block_reduce(data, (50, 50))
old: 139 ms    new: 14.6 ms    ratio: 9.52
>>> %timeit block_reduce(data, (100, 100))
old: 138 ms    new: 13.8 ms    ratio: 10.0
```

pllim commented:
Nice! Should we add this to https://github.com/astropy/astropy-benchmarks ?
